src/aspire/abinitio/commonline_sync3n.py

```diff
@@ -0,0 +1,1025 @@
+        Pmin = max(Pmin, 0)  # Clamp probability to [0,1]
+        Pmax = Pmax or 1
+        Pmax = min(Pmax, 1)  # Clamp probability to [0,1]
+
+        if scores_hist is None:
+            scores_hist = self._triangle_scores_inner(Rijs)
+
+        # Histogram decomposition: P & sigma evaluation
+        h = 1 / self.hist_intervals
+        hist_x = np.arange(h / 2, 1, h)
+        # normalization factor of one component of the histogram
+        A = (
+            (self.n_img * (self.n_img - 1) * (self.n_img - 2) / 2)
+            / self.hist_intervals
+            * (a + 1)
+        )
+        # normalization of 2nd component: B = P*N_delta/sum(f), where f is the component formula
+        B0 = P ** (self.n_img * (self.n_img - 1) * (self.n_img - 2) / 2) / np.sum(
+            ((1 - hist_x) ** b) * np.exp(-b / (1 - x0) * (1 - hist_x))
+        )
+        start_values = np.array([B0, P, b, x0], dtype=np.float64)
+        lower_bounds = np.array([0, Pmin**3, 2, 0], dtype=np.float64)
+        upper_bounds = np.array([np.inf, Pmax**3, np.inf, 1], dtype=np.float64)
+
+        # Fit distribution
+        def fun(x, B, P, b, x0, A=A, a=a):
+            """Function to fit. x is data vector."""
+            return (1 - P) * A * (1 - x) ** a + P * B * (1 - x) ** b * np.exp(
+                -b / (1 - x0) * (1 - x)
+            )
+
+        popt, pcov = curve_fit(
+            fun,
+            hist_x.astype(np.float64, copy=False),
+            scores_hist.astype(np.float64, copy=False),
+            p0=start_values,
+            bounds=(lower_bounds, upper_bounds),
+        )
+        B, P, b, x0 = popt
+
+        # Derive P and sigma
+        P = P ** (1 / 3)
+        peak = x0  # can rm later
+        sigma = (1 - peak) / peak2sigma
+
+        # Initialize probability computations
+        # Local histograms analysis
+        A = a + 1  # distribution 1st component normalization factor
+        # distribution 2nd component normalization factor
+        B = B / (
+            (self.n_img * (self.n_img - 1) * (self.n_img - 2) / 2) / self.hist_intervals
+        )
+
+        # Calculate probabilities
+        ln_f_ind, ln_f_arb = self._pairs_probabilities(Rijs, P**2, A, a, B, b, x0)
+        Pij = 1 / (1 + (1 - P) / P * np.exp(ln_f_arb - ln_f_ind))
+
+        # Fix singular output
+        num_nan = np.sum(np.isnan(Pij))
+        if num_nan > 0:
+            logger.error(
+                f"NaN probabilities occurred {num_nan} times out of {np.size(Pij)}. Setting NaNs to zero."
+            )
+            Pij = np.nan_to_num(Pij)
+
+        return P, sigma, Pij, scores_hist
+
+    ###########################################
+    # Primary Methods                         #
+    ###########################################
+
+    def _estimate_relative_viewing_directions(self):
+        """
+        Estimate the relative viewing directions vij = vi*vj^T, i<j, and vii = vi*vi^T, where
+        vi is the third row of the i'th rotation matrix Ri.
+        """
```

Comment:
May want to change this to `_estimate_relative_rotations()` (and the docstring) or possibly get rid of this method altogether and just call `_estimate_all_Rijs(self.clmatrix)` from the main `estimate_rotations()` method.

Reply:
Sure, I just moved the functional lines into `estimate_rotations`